File: orcastork/adapters/redis/store.py
"""Redis ``DataPointStore`` — keyed-merge over a hash, revision + epoch via a Lua CAS.

Each identity ``(type, value)`` is a field in ``dp:{session}`` holding the DataPoint JSON;
its first-added and last-freshened revisions live in parallel hashes (``added:{session}`` /
``upd:{session}``) keyed by the same identity. The session revision (``rev:{session}``) and
fencing epoch (``epoch:{session}``) are integer keys. The keyed-merge is computed in Python
(the orchestrator is the sole per-session mutator), then applied **atomically and
epoch-guarded** by a single Lua script: a stale epoch is rejected with no partial write, and
the revision is allocated by ``INCR`` **inside the script** so it stays monotonic regardless
of interleaving (the script treats the DataPoint JSON as an opaque blob, so it never has to
decode it). ``apply_resolved`` runs the same script without the ``write`` path's prior
HGETALL: a sole-mutator caller that already resolved the merge supplies the 'a'/'u' mode per
identity itself, so the hot write path costs one round-trip instead of a full re-read.
"""

from __future__ import annotations

from collections.abc import Iterable, Sequence
from datetime import datetime
from typing import Any

import orjson
from redis.asyncio import Redis

from ...datapoints import BaseDataPoint, DataPointView, canonical_value, parse_data_point
from ...exceptions import StaleEpochError
from ...ids import Epoch, OperatorId, Revision, SessionId
from ...ports.change_set import ChangeSet
from ...ports.datapoint_store import EffectClaim
from .ttl import DEFAULT_STATE_TTL_MS, slide_ttl

# Atomically: reject a stale epoch (no writes); else bump epoch, and if there are changes
# allocate the next revision with INCR and apply each field. KEYS = epoch, rev, dp, added,
# upd; ARGV = epoch, then triples (identity, mode, dp_json) where mode 'a' = new identity
# (stamp added + updated) and 'u' = re-observation (stamp updated only).
_WRITE_SCRIPT = """
local epoch = tonumber(ARGV[1])
local stored = tonumber(redis.call('GET', KEYS[1]) or '0')
if epoch < stored then return -1 end
if epoch > stored then redis.call('SET', KEYS[1], epoch) end
if #ARGV < 4 then return tonumber(redis.call('GET', KEYS[2]) or '0') end
local rev = redis.call('INCR', KEYS[2])
for i = 2, #ARGV, 3 do
  local identity = ARGV[i]
  redis.call('HSET', KEYS[3], identity, ARGV[i + 2])
  redis.call('HSET', KEYS[5], identity, rev)
  if ARGV[i + 1] == 'a' then redis.call('HSET', KEYS[4], identity, rev) end
end
return rev
"""

# Atomically epoch-guard a session-meta hash-field write (a watermark, the wall-clock deadline,
# the flow fingerprint): a stale epoch is rejected, and the highest ACCEPTED epoch is recorded.
# Ownership is the lock's job; recording the epoch on every guarded write is what completes
# stale-writer rejection — a session whose first store mutation is one of these fields must
# still fence a later lower-epoch writer.
_GUARDED_FIELD_SCRIPT = """
local epoch = tonumber(ARGV[1])
local stored = tonumber(redis.call('GET', KEYS[1]) or '0')
if epoch < stored then return -1 end
if epoch > stored then redis.call('SET', KEYS[1], epoch) end
redis.call('HSET', KEYS[2], ARGV[2], ARGV[3])
return 0
"""

# Atomically epoch-guard a side-effect claim, recording the highest accepted epoch exactly like
# every other guarded write (a session whose first mutation is an effect claim must still fence
# a later lower-epoch writer). The stored states are 'pending:<epoch>' (in-flight, naming its
# owner) and 'committed' (the effect ran). KEYS = epoch, fx hash; ARGV = epoch, effect key,
# reclaim_stale flag ('1'/'0'). Returns -1 on a stale epoch, else the EffectClaim code:
# 1 ACQUIRED, 2 ALREADY_COMMITTED, 3 PENDING_SAME_EPOCH, 4 PENDING_STALE_EPOCH.
_CLAIM_EFFECT_SCRIPT = """
local epoch = tonumber(ARGV[1])
local stored = tonumber(redis.call('GET', KEYS[1]) or '0')
if epoch < stored then return -1 end
if epoch > stored then redis.call('SET', KEYS[1], epoch) end
local state = redis.call('HGET', KEYS[2], ARGV[2])
local pending = 'pending:' .. ARGV[1]
if state == false then
  redis.call('HSET', KEYS[2], ARGV[2], pending)
  return 1
end
if state == 'committed' then return 2 end
if state == pending then return 3 end
if ARGV[3] == '1' then
  redis.call('HSET', KEYS[2], ARGV[2], pending)
  return 1
end
return 4
"""

# Atomically epoch-guard the pending → committed transition (recording the accepted epoch like
# every guarded write). Only this epoch's own pending mark transitions; 'committed' stays as-is
# (idempotent) and any other state is left untouched — the script never fabricates 'committed'
# for a claim this epoch does not own. -1 on a stale epoch.
_COMMIT_EFFECT_SCRIPT = """
local epoch = tonumber(ARGV[1])
local stored = tonumber(redis.call('GET', KEYS[1]) or '0')
if epoch < stored then return -1 end
if epoch > stored then redis.call('SET', KEYS[1], epoch) end
if redis.call('HGET', KEYS[2], ARGV[2]) == ('pending:' .. ARGV[1]) then
  redis.call('HSET', KEYS[2], ARGV[2], 'committed')
end
return 0
"""

# Atomically epoch-guard the revert (recording the accepted epoch like every guarded write):
# deletes ONLY this epoch's own pending mark — never 'committed' (the effect DID run) and never
# another epoch's pending. -1 on a stale epoch.
_REVERT_EFFECT_SCRIPT = """
local epoch = tonumber(ARGV[1])
local stored = tonumber(redis.call('GET', KEYS[1]) or '0')
if epoch < stored then return -1 end
if epoch > stored then redis.call('SET', KEYS[1], epoch) end
if redis.call('HGET', KEYS[2], ARGV[2]) == ('pending:' .. ARGV[1]) then
  redis.call('HDEL', KEYS[2], ARGV[2])
end
return 0
"""

_CLAIM_RESULTS = {
    1: EffectClaim.ACQUIRED,
    2: EffectClaim.ALREADY_COMMITTED,
    3: EffectClaim.PENDING_SAME_EPOCH,
    4: EffectClaim.PENDING_STALE_EPOCH,
}


class RedisDataPointStore:
    def __init__(self, redis: Redis, *, state_ttl_ms: int = DEFAULT_STATE_TTL_MS) -> None:
        # redis-py types its commands as a sync/async union (ResponseT); treat the client as
        # Any internally so awaits type cleanly, while the public param stays typed.
        self._redis: Any = redis
        self._state_ttl_ms = state_ttl_ms

    @staticmethod
    def _identity(data_point: BaseDataPoint[Any]) -> str:
        # ``canonical_value`` is the shared string form of the keyed-merge identity (same normalization
        # as the in-memory store and the archive), so dedup is identical across adapters.
        return f'{data_point.type}\x00{canonical_value(data_point.value)}'

    @staticmethod
    def _keys(session_id: SessionId) -> tuple[str, str, str, str, str]:
        return (
            f'epoch:{session_id}',
            f'rev:{session_id}',
            f'dp:{session_id}',
            f'added:{session_id}',
            f'upd:{session_id}',
        )

    async def write(
        self, session_id: SessionId, data_points: Iterable[BaseDataPoint[Any]], *, epoch: Epoch
    ) -> Revision:
        _, _, dp_key, _, _ = self._keys(session_id)
        current = await self._redis.hgetall(dp_key)
        triples: list[str] = []
        for data_point in data_points:
            identity = self._identity(data_point)
            existing_raw = current.get(identity)
            if existing_raw is None:
                # The client decodes responses, so values flow as str — decode the orjson bytes here.
                triples += [identity, 'a', orjson.dumps(data_point.model_dump(mode='json')).decode()]
            else:
                existing_dp = parse_data_point(orjson.loads(existing_raw))
                if data_point.last_retrieved > existing_dp.last_retrieved:
                    merged = existing_dp.reobserved(data_point.last_retrieved)
                    triples += [identity, 'u', orjson.dumps(merged.model_dump(mode='json')).decode()]
        return await self._apply(session_id, triples, epoch)

    async def apply_resolved(
        self,
        session_id: SessionId,
        *,
        added: Sequence[BaseDataPoint[Any]],
        updated: Sequence[BaseDataPoint[Any]],
        epoch: Epoch,
    ) -> Revision:
        # The sole-mutator fast path: the caller already keyed-merged, so the prior HGETALL that
        # `write` needs to decide 'a' vs 'u' is skipped entirely — the supplied split IS the mode.
        triples: list[str] = []
        for data_point in added:
            triples += [self._identity(data_point), 'a', orjson.dumps(data_point.model_dump(mode='json')).decode()]
        for data_point in updated:
            triples += [self._identity(data_point), 'u', orjson.dumps(data_point.model_dump(mode='json')).decode()]
        return await self._apply(session_id, triples, epoch)

    async def _apply(self, session_id: SessionId, triples: Sequence[str], epoch: Epoch) -> Revision:
        epoch_key, rev_key, dp_key, added_key, upd_key = self._keys(session_id)
        result = await self._redis.eval(
            _WRITE_SCRIPT, 5, epoch_key, rev_key, dp_key, added_key, upd_key, int(epoch), *triples
        )
        if int(result) == -1:
            raise StaleEpochError(f'epoch {epoch} is stale for session {session_id}')
        # Effect marks (`fx:`) and session meta (`meta:` — the wall-clock deadline and the flow
        # fingerprint) are slid here too, so they share the session's sliding lifetime: an active
        # session must never lose its at-most-once guarantees or its remaining deadline budget
        # mid-flight.
        await slide_ttl(
            self._redis,
            self._state_ttl_ms,
            epoch_key,
            rev_key,
            dp_key,
            added_key,
            upd_key,
            f'wm:{session_id}',
            f'fx:{session_id}',
            f'meta:{session_id}',
        )
        return Revision(int(result))

    async def _entries(self, session_id: SessionId) -> list[dict[str, Any]]:
        _, _, dp_key, added_key, upd_key = self._keys(session_id)
        dps = await self._redis.hgetall(dp_key)
        added = await self._redis.hgetall(added_key)
        updated = await self._redis.hgetall(upd_key)
        return [
            {'dp': orjson.loads(raw), 'added': int(added.get(identity, 0)), 'updated': int(updated.get(identity, 0))}
            for identity, raw in dps.items()
        ]

    async def snapshot(self, session_id: SessionId) -> DataPointView:
        return DataPointView(parse_data_point(entry['dp']) for entry in await self._entries(session_id))

    async def revision(self, session_id: SessionId) -> Revision:
        _, rev_key, _, _, _ = self._keys(session_id)
        return Revision(int(await self._redis.get(rev_key) or 0))

    async def change_set_since(self, session_id: SessionId, since: Revision) -> ChangeSet:
        added: list[BaseDataPoint[Any]] = []
        updated: list[BaseDataPoint[Any]] = []
        for entry in await self._entries(session_id):
            data_point = parse_data_point(entry['dp'])
            if entry['added'] > since:
                added.append(data_point)
            elif entry['updated'] > since:
                updated.append(data_point)
        return ChangeSet(added=tuple(added), updated=tuple(updated))

    async def get_watermark(self, session_id: SessionId, operator_id: OperatorId) -> Revision | None:
        raw = await self._redis.hget(f'wm:{session_id}', operator_id)
        return None if raw is None else Revision(int(raw))

    async def set_watermark(
        self, session_id: SessionId, operator_id: OperatorId, revision: Revision, *, epoch: Epoch
    ) -> None:
        epoch_key, _, _, _, _ = self._keys(session_id)
        result = await self._redis.eval(
            _GUARDED_FIELD_SCRIPT, 2, epoch_key, f'wm:{session_id}', str(int(epoch)), operator_id, str(int(revision))
        )
        if int(result) == -1:
            raise StaleEpochError(f'epoch {epoch} is stale for session {session_id}')
        await slide_ttl(self._redis, self._state_ttl_ms, epoch_key, f'wm:{session_id}')

    async def claim_effect(
        self, session_id: SessionId, effect_key: str, *, epoch: Epoch, reclaim_stale: bool
    ) -> EffectClaim:
        result = await self._run_effect_script(
            _CLAIM_EFFECT_SCRIPT, session_id, effect_key, epoch, '1' if reclaim_stale else '0'
        )
        return _CLAIM_RESULTS[result]

    async def commit_effect(self, session_id: SessionId, effect_key: str, *, epoch: Epoch) -> None:
        await self._run_effect_script(_COMMIT_EFFECT_SCRIPT, session_id, effect_key, epoch)

    async def revert_effect(self, session_id: SessionId, effect_key: str, *, epoch: Epoch) -> None:
        await self._run_effect_script(_REVERT_EFFECT_SCRIPT, session_id, effect_key, epoch)

    async def get_effect_state(self, session_id: SessionId, effect_key: str) -> str | None:
        raw = await self._redis.hget(f'fx:{session_id}', effect_key)
        return None if raw is None else str(raw)

    async def _run_effect_script(
        self, script: str, session_id: SessionId, effect_key: str, epoch: Epoch, *extra_args: str
    ) -> int:
        epoch_key, _, _, _, _ = self._keys(session_id)
        result = await self._redis.eval(
            script, 2, epoch_key, f'fx:{session_id}', str(int(epoch)), effect_key, *extra_args
        )
        if int(result) == -1:
            raise StaleEpochError(f'epoch {epoch} is stale for session {session_id}')
        await slide_ttl(self._redis, self._state_ttl_ms, epoch_key, f'fx:{session_id}')
        return int(result)

    async def get_session_deadline(self, session_id: SessionId) -> datetime | None:
        raw = await self._redis.hget(f'meta:{session_id}', 'deadline')
        return None if raw is None else datetime.fromisoformat(raw)

    async def set_session_deadline(self, session_id: SessionId, deadline: datetime, *, epoch: Epoch) -> None:
        epoch_key, _, _, _, _ = self._keys(session_id)
        result = await self._redis.eval(
            _GUARDED_FIELD_SCRIPT,
            2,
            epoch_key,
            f'meta:{session_id}',
            str(int(epoch)),
            'deadline',
            deadline.isoformat(),
        )
        if int(result) == -1:
            raise StaleEpochError(f'epoch {epoch} is stale for session {session_id}')
        await slide_ttl(self._redis, self._state_ttl_ms, epoch_key, f'meta:{session_id}')

    async def get_flow_fingerprint(self, session_id: SessionId) -> str | None:
        raw = await self._redis.hget(f'meta:{session_id}', 'flow_fingerprint')
        return None if raw is None else str(raw)

    async def set_flow_fingerprint(self, session_id: SessionId, fingerprint: str, *, epoch: Epoch) -> None:
        epoch_key, _, _, _, _ = self._keys(session_id)
        result = await self._redis.eval(
            _GUARDED_FIELD_SCRIPT,
            2,
            epoch_key,
            f'meta:{session_id}',
            str(int(epoch)),
            'flow_fingerprint',
            fingerprint,
        )
        if int(result) == -1:
            raise StaleEpochError(f'epoch {epoch} is stale for session {session_id}')
        await slide_ttl(self._redis, self._state_ttl_ms, epoch_key, f'meta:{session_id}')

File: typescript/src/orcastork/adapters/redis/store.ts
/**
 * Redis `DataPointStore` — keyed-merge over a hash, revision + epoch via a Lua CAS.
 *
 * Each identity `(type, value)` is a field in `dp:{session}` holding the DataPoint JSON; its
 * first-added and last-freshened revisions live in parallel hashes (`added:{session}` /
 * `upd:{session}`) keyed by the same identity. The session revision (`rev:{session}`) and fencing
 * epoch (`epoch:{session}`) are integer keys. The keyed-merge is computed here (the orchestrator is
 * the sole per-session mutator), then applied **atomically and epoch-guarded** by a single Lua
 * script: a stale epoch is rejected with no partial write, and the revision is allocated by `INCR`
 * **inside the script** so it stays monotonic regardless of interleaving (the script treats the
 * DataPoint JSON as an opaque blob, so it never has to decode it). `applyResolved` runs the same
 * script without the `write` path's prior HGETALL: a sole-mutator caller that already resolved the
 * merge supplies the 'a'/'u' mode per identity itself, so the hot write path costs one round-trip
 * instead of a full re-read.
 *
 * @module
 */

import type { AnyDataPoint } from '../../datapoints/index.js';
import { DataPointView, identityKey, parseDataPoint } from '../../datapoints/index.js';
import { OrchestrationError, StaleEpochError } from '../../exceptions.js';
import type { Epoch, OperatorId, Revision, SessionId } from '../../ids.js';
import { Revision as toRevision } from '../../ids.js';
import { ChangeSet } from '../../ports/change_set.js';
import type { ApplyResolvedOptions, ClaimEffectOptions, DataPointStore } from '../../ports/datapoint_store.js';
import { EffectClaim } from '../../ports/datapoint_store.js';
import type { RedisEvalCommand, RedisExpireCommand } from './ttl.js';
import { DEFAULT_STATE_TTL_MS, slideTtl } from './ttl.js';

/** The commands this store needs. */
export interface RedisStoreClient extends RedisEvalCommand, RedisExpireCommand {
  get(key: string): Promise<string | null>;
  hGet(key: string, field: string): Promise<string | null>;
  hGetAll(key: string): Promise<Record<string, string>>;
}

// Atomically: reject a stale epoch (no writes); else bump epoch, and if there are changes
// allocate the next revision with INCR and apply each field. KEYS = epoch, rev, dp, added,
// upd; ARGV = epoch, then triples (identity, mode, dp_json) where mode 'a' = new identity
// (stamp added + updated) and 'u' = re-observation (stamp updated only).
const WRITE_SCRIPT = `
local epoch = tonumber(ARGV[1])
local stored = tonumber(redis.call('GET', KEYS[1]) or '0')
if epoch < stored then return -1 end
if epoch > stored then redis.call('SET', KEYS[1], epoch) end
if #ARGV < 4 then return tonumber(redis.call('GET', KEYS[2]) or '0') end
local rev = redis.call('INCR', KEYS[2])
for i = 2, #ARGV, 3 do
  local identity = ARGV[i]
  redis.call('HSET', KEYS[3], identity, ARGV[i + 2])
  redis.call('HSET', KEYS[5], identity, rev)
  if ARGV[i + 1] == 'a' then redis.call('HSET', KEYS[4], identity, rev) end
end
return rev
`;

// Atomically epoch-guard a session-meta hash-field write (a watermark, the wall-clock deadline,
// the flow fingerprint): a stale epoch is rejected, and the highest ACCEPTED epoch is recorded.
// Ownership is the lock's job; recording the epoch on every guarded write is what completes
// stale-writer rejection — a session whose first store mutation is one of these fields must
// still fence a later lower-epoch writer.
const GUARDED_FIELD_SCRIPT = `
local epoch = tonumber(ARGV[1])
local stored = tonumber(redis.call('GET', KEYS[1]) or '0')
if epoch < stored then return -1 end
if epoch > stored then redis.call('SET', KEYS[1], epoch) end
redis.call('HSET', KEYS[2], ARGV[2], ARGV[3])
return 0
`;

// Atomically epoch-guard a side-effect claim, recording the highest accepted epoch exactly like
// every other guarded write (a session whose first mutation is an effect claim must still fence
// a later lower-epoch writer). The stored states are 'pending:<epoch>' (in-flight, naming its
// owner) and 'committed' (the effect ran). KEYS = epoch, fx hash; ARGV = epoch, effect key,
// reclaim_stale flag ('1'/'0'). Returns -1 on a stale epoch, else the EffectClaim code:
// 1 ACQUIRED, 2 ALREADY_COMMITTED, 3 PENDING_SAME_EPOCH, 4 PENDING_STALE_EPOCH.
const CLAIM_EFFECT_SCRIPT = `
local epoch = tonumber(ARGV[1])
local stored = tonumber(redis.call('GET', KEYS[1]) or '0')
if epoch < stored then return -1 end
if epoch > stored then redis.call('SET', KEYS[1], epoch) end
local state = redis.call('HGET', KEYS[2], ARGV[2])
local pending = 'pending:' .. ARGV[1]
if state == false then
  redis.call('HSET', KEYS[2], ARGV[2], pending)
  return 1
end
if state == 'committed' then return 2 end
if state == pending then return 3 end
if ARGV[3] == '1' then
  redis.call('HSET', KEYS[2], ARGV[2], pending)
  return 1
end
return 4
`;

// Atomically epoch-guard the pending → committed transition (recording the accepted epoch like
// every guarded write). Only this epoch's own pending mark transitions; 'committed' stays as-is
// (idempotent) and any other state is left untouched — the script never fabricates 'committed'
// for a claim this epoch does not own. -1 on a stale epoch.
const COMMIT_EFFECT_SCRIPT = `
local epoch = tonumber(ARGV[1])
local stored = tonumber(redis.call('GET', KEYS[1]) or '0')
if epoch < stored then return -1 end
if epoch > stored then redis.call('SET', KEYS[1], epoch) end
if redis.call('HGET', KEYS[2], ARGV[2]) == ('pending:' .. ARGV[1]) then
  redis.call('HSET', KEYS[2], ARGV[2], 'committed')
end
return 0
`;

// Atomically epoch-guard the revert (recording the accepted epoch like every guarded write):
// deletes ONLY this epoch's own pending mark — never 'committed' (the effect DID run) and never
// another epoch's pending. -1 on a stale epoch.
const REVERT_EFFECT_SCRIPT = `
local epoch = tonumber(ARGV[1])
local stored = tonumber(redis.call('GET', KEYS[1]) or '0')
if epoch < stored then return -1 end
if epoch > stored then redis.call('SET', KEYS[1], epoch) end
if redis.call('HGET', KEYS[2], ARGV[2]) == ('pending:' .. ARGV[1]) then
  redis.call('HDEL', KEYS[2], ARGV[2])
end
return 0
`;

/** The claim script's integer code, as the port's outcome. */
const claimResult = (code: number): EffectClaim => {
  switch (code) {
    case 1:
      return EffectClaim.ACQUIRED;
    case 2:
      return EffectClaim.ALREADY_COMMITTED;
    case 3:
      return EffectClaim.PENDING_SAME_EPOCH;
    case 4:
      return EffectClaim.PENDING_STALE_EPOCH;
    default:
      // Unreachable while the script and this table agree; a mismatch is a deployment running two
      // versions of the adapter against one keyspace, which must fail loudly rather than guess.
      throw new OrchestrationError(`the effect claim script returned an unknown code ${code}`);
  }
};

/** The five per-session keys every write path touches, in the order the scripts expect them. */
interface SessionKeys {
  readonly epoch: string;
  readonly revision: string;
  readonly dataPoints: string;
  readonly added: string;
  readonly updated: string;
}

/** One stored identity, as `_entries` reads it back. */
interface StoredEntry {
  readonly dataPoint: AnyDataPoint;
  readonly added: number;
  readonly updated: number;
}

/** How a {@link RedisDataPointStore} is configured. */
export interface RedisDataPointStoreOptions {
  /** The sliding lifetime of the session's Redis state. */
  readonly stateTtlMs?: number;
}

/** The session blackboard, held in Redis and fenced by a Lua CAS. */
export class RedisDataPointStore implements DataPointStore {
  private readonly redis: RedisStoreClient;
  private readonly stateTtlMs: number;

  public constructor(redis: RedisStoreClient, options: RedisDataPointStoreOptions = {}) {
    this.redis = redis;
    this.stateTtlMs = options.stateTtlMs ?? DEFAULT_STATE_TTL_MS;
  }

  /**
   * The keyed-merge identity of a DataPoint, as a Redis hash field.
   *
   * `identityKey` is the shared string form (the same normalization as the in-memory store and the
   * archive), so dedup is identical across adapters — and across runtimes.
   */
  private static identity(dataPoint: AnyDataPoint): string {
    return identityKey(dataPoint);
  }

  private static keys(sessionId: SessionId): SessionKeys {
    return {
      epoch: `epoch:${sessionId}`,
      revision: `rev:${sessionId}`,
      dataPoints: `dp:${sessionId}`,
      added: `added:${sessionId}`,
      updated: `upd:${sessionId}`,
    };
  }

  public async write(
    sessionId: SessionId,
    dataPoints: Iterable<AnyDataPoint>,
    options: { readonly epoch: Epoch },
  ): Promise<Revision> {
    const current = await this.redis.hGetAll(RedisDataPointStore.keys(sessionId).dataPoints);
    const triples: string[] = [];
    for (const dataPoint of dataPoints) {
      const identity = RedisDataPointStore.identity(dataPoint);
      const existingRaw = current[identity];
      if (existingRaw === undefined) {
        triples.push(identity, 'a', JSON.stringify(dataPoint.toWire()));
        continue;
      }
      const existing = parseDataPoint(JSON.parse(existingRaw));
      if (dataPoint.lastRetrieved.getTime() > existing.lastRetrieved.getTime()) {
        const merged = existing.reobserved(dataPoint.lastRetrieved);
        triples.push(identity, 'u', JSON.stringify(merged.toWire()));
      }
    }
    return this.apply(sessionId, triples, options.epoch);
  }

  public async applyResolved(sessionId: SessionId, options: ApplyResolvedOptions): Promise<Revision> {
    // The sole-mutator fast path: the caller already keyed-merged, so the prior HGETALL that
    // `write` needs to decide 'a' vs 'u' is skipped entirely — the supplied split IS the mode.
    const triples: string[] = [];
    for (const dataPoint of options.added) {
      triples.push(RedisDataPointStore.identity(dataPoint), 'a', JSON.stringify(dataPoint.toWire()));
    }
    for (const dataPoint of options.updated) {
      triples.push(RedisDataPointStore.identity(dataPoint), 'u', JSON.stringify(dataPoint.toWire()));
    }
    return this.apply(sessionId, triples, options.epoch);
  }

  private async apply(sessionId: SessionId, triples: readonly string[], epoch: Epoch): Promise<Revision> {
    const keys = RedisDataPointStore.keys(sessionId);
    const result = Number(
      await this.redis.eval(WRITE_SCRIPT, {
        keys: [keys.epoch, keys.revision, keys.dataPoints, keys.added, keys.updated],
        arguments: [String(epoch), ...triples],
      }),
    );
    if (result === -1) {
      throw new StaleEpochError(`epoch ${epoch} is stale for session ${sessionId}`);
    }
    // Effect marks (`fx:`) and session meta (`meta:` — the wall-clock deadline and the flow
    // fingerprint) are slid here too, so they share the session's sliding lifetime: an active
    // session must never lose its at-most-once guarantees or its remaining deadline budget
    // mid-flight.
    await slideTtl(
      this.redis,
      this.stateTtlMs,
      keys.epoch,
      keys.revision,
      keys.dataPoints,
      keys.added,
      keys.updated,
      `wm:${sessionId}`,
      `fx:${sessionId}`,
      `meta:${sessionId}`,
    );
    return toRevision(result);
  }

  private async entries(sessionId: SessionId): Promise<readonly StoredEntry[]> {
    const keys = RedisDataPointStore.keys(sessionId);
    const dataPoints = await this.redis.hGetAll(keys.dataPoints);
    const added = await this.redis.hGetAll(keys.added);
    const updated = await this.redis.hGetAll(keys.updated);
    return Object.entries(dataPoints).map(([identity, raw]) => ({
      dataPoint: parseDataPoint(JSON.parse(raw)),
      added: Number(added[identity] ?? 0),
      updated: Number(updated[identity] ?? 0),
    }));
  }

  public async snapshot(sessionId: SessionId): Promise<DataPointView> {
    return new DataPointView((await this.entries(sessionId)).map((entry) => entry.dataPoint));
  }

  public async revision(sessionId: SessionId): Promise<Revision> {
    const stored = await this.redis.get(RedisDataPointStore.keys(sessionId).revision);
    return toRevision(stored === null ? 0 : Number(stored));
  }

  public async changeSetSince(sessionId: SessionId, since: Revision): Promise<ChangeSet> {
    const added: AnyDataPoint[] = [];
    const updated: AnyDataPoint[] = [];
    for (const entry of await this.entries(sessionId)) {
      if (entry.added > since) {
        added.push(entry.dataPoint);
      } else if (entry.updated > since) {
        updated.push(entry.dataPoint);
      }
    }
    return new ChangeSet({ added, updated });
  }

  public async getWatermark(sessionId: SessionId, operatorId: OperatorId): Promise<Revision | null> {
    const raw = await this.redis.hGet(`wm:${sessionId}`, operatorId);
    return raw === null ? null : toRevision(Number(raw));
  }

  public async setWatermark(
    sessionId: SessionId,
    operatorId: OperatorId,
    revision: Revision,
    options: { readonly epoch: Epoch },
  ): Promise<void> {
    await this.guardedField(sessionId, `wm:${sessionId}`, operatorId, String(revision), options.epoch);
  }

  public async claimEffect(
    sessionId: SessionId,
    effectKey: string,
    options: ClaimEffectOptions,
  ): Promise<EffectClaim> {
    return claimResult(
      await this.runEffectScript(CLAIM_EFFECT_SCRIPT, sessionId, effectKey, options.epoch, [
        options.reclaimStale ? '1' : '0',
      ]),
    );
  }

  public async commitEffect(
    sessionId: SessionId,
    effectKey: string,
    options: { readonly epoch: Epoch },
  ): Promise<void> {
    await this.runEffectScript(COMMIT_EFFECT_SCRIPT, sessionId, effectKey, options.epoch, []);
  }

  public async revertEffect(
    sessionId: SessionId,
    effectKey: string,
    options: { readonly epoch: Epoch },
  ): Promise<void> {
    await this.runEffectScript(REVERT_EFFECT_SCRIPT, sessionId, effectKey, options.epoch, []);
  }

  public async getEffectState(sessionId: SessionId, effectKey: string): Promise<string | null> {
    return this.redis.hGet(`fx:${sessionId}`, effectKey);
  }

  private async runEffectScript(
    script: string,
    sessionId: SessionId,
    effectKey: string,
    epoch: Epoch,
    extraArguments: readonly string[],
  ): Promise<number> {
    const epochKey = RedisDataPointStore.keys(sessionId).epoch;
    const result = Number(
      await this.redis.eval(script, {
        keys: [epochKey, `fx:${sessionId}`],
        arguments: [String(epoch), effectKey, ...extraArguments],
      }),
    );
    if (result === -1) {
      throw new StaleEpochError(`epoch ${epoch} is stale for session ${sessionId}`);
    }
    await slideTtl(this.redis, this.stateTtlMs, epochKey, `fx:${sessionId}`);
    return result;
  }

  public async getSessionDeadline(sessionId: SessionId): Promise<Date | null> {
    const raw = await this.redis.hGet(`meta:${sessionId}`, 'deadline');
    return raw === null ? null : new Date(raw);
  }

  public async setSessionDeadline(
    sessionId: SessionId,
    deadline: Date,
    options: { readonly epoch: Epoch },
  ): Promise<void> {
    await this.guardedField(sessionId, `meta:${sessionId}`, 'deadline', deadline.toISOString(), options.epoch);
  }

  public async getFlowFingerprint(sessionId: SessionId): Promise<string | null> {
    return this.redis.hGet(`meta:${sessionId}`, 'flow_fingerprint');
  }

  public async setFlowFingerprint(
    sessionId: SessionId,
    fingerprint: string,
    options: { readonly epoch: Epoch },
  ): Promise<void> {
    await this.guardedField(sessionId, `meta:${sessionId}`, 'flow_fingerprint', fingerprint, options.epoch);
  }

  /** One epoch-guarded hash-field write, plus the TTL slide that keeps the field alive with it. */
  private async guardedField(
    sessionId: SessionId,
    hashKey: string,
    field: string,
    value: string,
    epoch: Epoch,
  ): Promise<void> {
    const epochKey = RedisDataPointStore.keys(sessionId).epoch;
    const result = Number(
      await this.redis.eval(GUARDED_FIELD_SCRIPT, {
        keys: [epochKey, hashKey],
        arguments: [String(epoch), field, value],
      }),
    );
    if (result === -1) {
      throw new StaleEpochError(`epoch ${epoch} is stale for session ${sessionId}`);
    }
    await slideTtl(this.redis, this.stateTtlMs, epochKey, hashKey);
  }
}
